src/include/optimizer/absexpr_expression.h

```diff
@@ -0,0 +1,183 @@
+  }
+
+  // Operator contains Physical node
+  bool IsPhysical() const {
+    return false;
+  }
+
+  std::string GetName() const {
+    if (IsDefined()) {
+      return node->GetExpressionName();
+    }
+
+    return "Undefined";
+  }
+
+  hash_t Hash() const {
+    if (IsDefined()) {
+      return node->Hash();
+    }
+    return 0;
+  }
+
+  bool operator==(const AbsExpr_Container &r) {
+    if (IsDefined() && r.IsDefined()) {
+      // (TODO): need a better way to determine deep equality
+
+      // NOTE:
+      // Without proper equality determinations, the groups will
+      // not be assigned correctly. Arguably, terrier does this
+      // better because a blind ExactlyEquals on different types
+      // of ConstantValueExpression under Peloton will crash!
+
+      // For now, just return (false).
+      // I don't anticipate this will affect correctness, just
+      // performance, since duplicate trees will have to evaluated
+      // over and over again, rather than being able to "borrow"
+      // a previous tree's rewrite.
+      //
+      // Probably not worth to create a "validator" since porting
+      // this to terrier anyways (?). == does not check Value
+      // so it's broken. ExactlyEqual requires precondition checking.
+      return false;
+    } else if (!IsDefined() && !r.IsDefined()) {
+      return true;
+    }
+    return false;
+  }
+
+  // Operator contains physical or logical operator node
+  bool IsDefined() const {
+    return node != nullptr;
+  }
+
+  //(TODO): fix memory management once go to terrier
+  expression::AbstractExpression *Rebuild(std::vector<expression::AbstractExpression*> children) {
+    switch (GetType()) {
+      case ExpressionType::COMPARE_EQUAL:
+      case ExpressionType::COMPARE_NOTEQUAL:
+      case ExpressionType::COMPARE_LESSTHAN:
+      case ExpressionType::COMPARE_GREATERTHAN:
+      case ExpressionType::COMPARE_LESSTHANOREQUALTO:
+      case ExpressionType::COMPARE_GREATERTHANOREQUALTO:
+      case ExpressionType::COMPARE_LIKE:
+      case ExpressionType::COMPARE_NOTLIKE:
+      case ExpressionType::COMPARE_IN:
+      case ExpressionType::COMPARE_DISTINCT_FROM: {
+        PELOTON_ASSERT(children.size() == 2);
+        return new expression::ComparisonExpression(GetType(), children[0], children[1]);
+      }
+      case ExpressionType::CONJUNCTION_AND:
+      case ExpressionType::CONJUNCTION_OR: {
+        PELOTON_ASSERT(children.size() == 2);
+        return new expression::ConjunctionExpression(GetType(), children[0], children[1]);
+      }
+      case ExpressionType::VALUE_CONSTANT: {
+        PELOTON_ASSERT(children.size() == 0);
+        auto cve = static_cast<const expression::ConstantValueExpression*>(node);
+        return new expression::ConstantValueExpression(cve->GetValue());
+      }
+      default: {
+        int type = static_cast<int>(GetType());
+        LOG_ERROR("Unimplemented Rebuild() for %d found", type);
+        return nullptr;
+      }
+    }
+  }
+
```

Comment:
What does Rebuild do in context of its AbsExpr_Container? You could probably add some more documentation for this function.